### DISCUSSION TITLE
IM overlay PNG over background JPG - windows preview and .bat file problems.

### DISCUSSION THREAD

**Discussion by Trailboy** [ImageMagick/ImageMagick#4397] [Digital Image Processing] [answered]
Hi all.

Firstly, using IM on win 7 with ...
![magick-Version](https://user-images.githubusercontent.com/64076927/138282885-8ebc88a6-5d73-4bb1-a034-da52744c35a2.JPG)
 
Files are here:
https://drive.google.com/drive/folders/12o4qifV2hWMfjJgrgwM1wqcIrWwD4ngu?usp=sharing

I have multiple .png files that I need to overlay on a Background.jpg image.

Using the code 

`magick composite img.png Background.jpg Img-overlay-Background.jpg`

works fine, except,

**Issue 1:**

the windows thumbnail preview files refuses to show the correct overlay image - it excludes the foreground.  The file itself is fine when opened in windows photo viewer.  Is this a problem inherent with windows preview thumbnails or is there something I can do about it?  Little disconcerting when you have to ensure that hundreds of images are overlaid correctly.  

![overlay-01](https://user-images.githubusercontent.com/64076927/138284386-32cdaa09-368b-433c-9647-ace11503c9b9.JPG)

**Issue 2:**

Trying to incorporate this into a windows .bat file so that I can convert all images in a folder in one go. 

The following .bat code syntax has worked for me previously (thanks to the [invaluable help ](https://github.com/ImageMagick/ImageMagick/discussions/4054)I always receive here, that is):

```
for %%F in (*.png) do magick %%F ^
	composite Background.jpg ^
	%%~nF_Background.jpg
```
but doesn't work here and complains about being unable to open image 'composite' rather than treating it as a command.  Is this command able to be incorporated into a .bat file?  

If anybody would be able to help this idiot out (yet again), I would be very grateful.

Thanks

**Comment by snibgo** [comment, marked as answer]
Issue 2 is easy. You have changed the order of the "magick" command. The word "composite" must be immediately after "magick" (as at the top of your post), not between the two filenames.

Issue 1 is a function of your operating system. I suspect you created a version of Img-overlay-Background.jpg without the foreground, and Windows showed that image. Then you re-created Img-overlay-Background.jpg with the foreground and Windows 7 didn't notice the change, and still showed the old thumbnail.

If I remember correctly, Win7 has a "refresh" button, possibly top-right. Hit that to re-create all thumbnails.

**Reply by Trailboy** [reply]
Hi Snibgo

Thanks for your valuable time!

Issue 1 - I know what you mean, but I have tried the refresh thumbnails, and also changed to View>>Medium Icons and then back to Large Icons (which usually works with Win7), but doesn't with this Imagemagick composite output.  Downloading the file img-overlay-background.jpg in the Drive link above (i.e. [this image](https://drive.google.com/file/d/1-FjfmggP2CQHQxihutAH-5eUhD9a4-qt/view?usp=sharing)) to my Win10 computer still shows the same problem - thumbnail doesn't reflect image correctly.  

You're right though, it's an OS problem.  I think issue 2 is probably more important to me anyway

Given your advice, this is the current code and it works fine save for the thumbnail preview!  

Thank you so much again.

```
for %%F in (*.png) do magick composite %%F ^
	Background.jpg ^
	%%~nF_Background.jpg
```
![Snigo-Issues](https://user-images.githubusercontent.com/64076927/138328230-82d71cc1-bff8-433b-887a-7202204cd132.JPG)

**Comment by snibgo** [comment]
Your Img-overlay-Background.jpg contains a thumbnail image. I suspect Win7 is using that for its thumbnail. In your magick command, you can insert "-strip" before writing the file.

However, "-strip" removes loads of metadata that you might prefer to leave unstripped, such as an ICC profile. A more targetted method uses exiftool to remove all tags that start with "Thumb":
```
exiftool -Thumb*= Img-overlay-Background.jpg
```
This creates a backup version of the old file named "Img-overlay-Background.jpg_original".

**Comment by Trailboy** [comment]
Hello again.  

Have a further issue for the fantastic brains trust that is this forum.  

Feel free to tell me where to go!

Previously, I have made sure that the transparent foreground.png and background.jpg image dimensions have always matched for very easy overlay composites. Both images were always in 'portrait' orientation.  

However, if I use a 'landscape' orientation foreground.png image, the output forces a match to the background.jpg and clips the right half of the foreground image, rather than prioritise the image dimensions of the foreground.png and force the background to match (hopefully clip the top and bottom so that the centre portion of the background image shows).   

Is it possible to code with composite so that the command retains the first foreground image dimensions rather than the second background image for its output?  The screenshot below shows ImgLandscape_Background.jpg as the code currently produces if it encounters a landscape image, but ideally I would like the command to produce the ImgLandscapeCorrect.jpg automatically.  

Is this possible? 

Images uploaded to the google drive link at the start of this thread.  

Thanks very much indeed to anybody who is able to help. 



![ImgLandscapePortraitProblem](https://user-images.githubusercontent.com/64076927/138442184-9fd968ce-a7bc-4011-b2db-221d6f7fed39.JPG)

**Comment by snibgo** [comment]
First point: I suggest not using "magick composite". Instead, use "magick" with "-composite". Yeah, I know that sounds strange.

You currently do this:
```
magick composite person.jpg Background.jpg out.jpg
```
Instead, I suggest this:
```
magick Background.jpg person.jpg -composite out.jpg
```
Note that the inputs are given in the opposite order.

Your Background.jpg is 2400x3600, so only 2400 pixels wide.

Your ImgLandscape.png is 3600x2400, so wider than the background. How do you want to deal with this? You say "hopefully clip the top and bottom [of the background] so that the centre portion of the background image shows". But this doesn't cope with the width problem.

A possible solution is to resize the background so it is exactly the same size as the foreground. To do this, we read the foreground first to get the size, read the background and resize it, then composite. Windows BAT script:
```
magick ^
  ImgLandscape.png ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( Background.jpg -resize "%%[MYSIZE]^!" ) ^
  +swap ^
  -composite ^
  out.jpg
```
If this is used inside a "for (...) loop, you need to escape the parentheses:
```
magick ^
  ImgLandscape.png ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ^( Background.jpg -resize "%%[MYSIZE]^!" ^) ^
  +swap ^
  -composite ^
  out.jpg
```
This does change the aspect ratio (ie width divided by height) of the background. I like this, but you might not. Alternatives include trimming the top or bottom or both, and resizing the remainder without changing the aspect ratio.
![out](https://user-images.githubusercontent.com/33812368/138451784-2850393d-7fa6-43ef-90ef-ed53a53ed8d9.jpg)

**Reply by Trailboy** [reply]
Thank you so much for the code Snibgo.  

Yes, all foreground images are sized 3600x2400px as we print no larger than 12x8 inch at 300dpi so doesn't stress our print-flow any more than needs be.  For convenience, we also size our background images to match, and accept that only the centre slice of the background will appear on the final image.  

Could I be cheeky and ask for the code that does not change the background aspect ratio, i.e. one that trims the background top and bottom.  As long as it isn't too time consuming for you, that is Snibgo!  You've already been too generous with your time.

The reason for asking is that the 'stretching' of the background to fit the foreground works great with a plain background (such as the blue above or the grey vignette first image below) and is actually preferable and looks really nice, a 'real world' background image such as the library below appears too distorted (second image) and isn't how the customer actually views their images before purchase (third image) using our online gallery.  

![BackgroundVignette](https://user-images.githubusercontent.com/64076927/138501625-fb27e22a-ad8a-4cb8-9a92-67ab9f904660.JPG)

![BackgroundLibrary](https://user-images.githubusercontent.com/64076927/138501675-0355bfeb-6db9-422a-8ff5-1cac41052912.JPG)

![BackgroundLibraryOnline](https://user-images.githubusercontent.com/64076927/138501816-00223999-794a-4785-98ec-636e31affc9f.JPG)

Still feel free to tell me where to go!

**Comment by snibgo** [comment]
This trims the top and bottom equally. If you instead want to trim just the top or just the bottom, change "-gravity Center" to "-gravity South" or "-gravity North".
```
magick ^
  ImgLandscape.png ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( Background.jpg -resize "%%[MYSIZE]^^" -gravity Center -crop "%%[MYSIZE]+0+0" +repage ) ^
  +swap ^
  -composite ^
  out2.jpg
```

**Comment by Trailboy** [comment]
Snibgo, thank you.

How can I show my gratitude?  [Donate ](https://imagemagick.org/script/support.php)to Development?

**Reply by snibgo** [reply]
Sure.
